File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.cs
public class Solution {
    public int MctFromLeafValues(int[] arr) {
        int[,] dp = new int[arr.Length,arr.Length];
        for(int i=arr.Length-1; i>=0; i--)
        {
            for(int j=0; j<arr.Length; j++)
            {
                if(i >= j)
                {
                    continue;
                }
                int min = Int32.MaxValue;
                for(int k=i; k<j; k++)
                {
                    int left = getMax(i,k,arr);
                    int right = getMax(k+1,j,arr);
                    min = Math.Min(min, left*right + dp[i,k] + dp[k+1,j]);
                }
                dp[i,j] = min;
            }
        }
        return dp[0,arr.Length-1];
    }
    // public int getMin(int i, int j, int[] arr, int[,] dp)
    // {
    //     // only one node can not lead to partitions
    //     if(i==j)
    //     {
    //         return 0;
    //     }
    //     if(dp[i,j] != -1)
    //     {
    //         return dp[i,j];
    //     }
    //     int min = Int32.MaxValue;
    //     for(int k=i; k<j; k++)
    //     {
    //         int left = getMax(i,k,arr);
    //         int right = getMax(k+1,j,arr);
    //         min = Math.Min(min, left*right + getMin(i,k,arr,dp) + getMin(k+1,j,arr,dp));
    //     }
    //     dp[i,j] = min;
    //     return min;
    // }
    private int getMax(int i, int j, int[] arr)
    {
        int max = -1;
        for(int k=i; k<=j; k++)
        {
            max = Math.Max(max,arr[k]);
        }
        return max;
    }
}

File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.java
class Solution {
    public int mctFromLeafValues(int[] arr) {
        int[][] dp = new int[arr.length][arr.length];
        for(int[] row : dp)
        {
            Arrays.fill(row, -1);
        }
        int ans = getAns(0,arr.length-1,arr, dp);
        return ans;
    }
    private int getAns(int i, int j, int[] arr, int[][] dp)
    {
        if(i >= j)
        {
            return 0;
        }
        if(dp[i][j] != -1)
        {
            return dp[i][j];
        }
        int ans = Integer.MAX_VALUE;
        for(int k=i; k<j; k++)
        {
            int maxLeft = getMax(i, k, arr);
            int maxRight = getMax(k+1, j, arr);
            ans = Math.min(ans, maxLeft*maxRight + getAns(i, k, arr, dp) + getAns(k+1, j, arr, dp));
        }
        return dp[i][j] = ans;
    }
    private int getMax(int i, int j, int[] arr)
    {
        int max = 0;
        for(int it=i; it<=j; it++)
        {
            max=Math.max(arr[it], max);
        }
        return max;
    }
}